src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
+  }
+  AstExpr(AstNodePtr ptr) : ptr{move(ptr)} {
+  }
+  AstExpr(const AstExpr& other) {
+    ptr = move(other.ptr);
+  }
+  AstExpr& operator=(const AstExpr& other) {
+    ptr = move(other.ptr);
+    return *this;
+  }
+  operator AstNodePtr() {
+    return move(ptr);
+  }
+  const AstNode* operator->() const {
+    return ptr.get();
+  }
+  mutable AstNodePtr ptr;
+};
+
+template <typename T, typename... Ts> AstExpr MakeExpr(Ts... ts) {
+  AstNode* node = dynamic_cast<AstNode*>(new T(std::forward<Ts>(ts)...));
+  CHECK(node);
+  return std::unique_ptr<AstNode>{node};
+}
+
+// AST term node, matches only if input contains term.
+class AstTermNode : public AstNode {
+ public:
+  AstTermNode(std::string term) : term_{move(term)} {
+  }
+  virtual bool Check(std::string_view input) const;
+  virtual std::string Debug() const;
+
+ private:
+  std::string term_;
+};
+
+// Ast negation node, matches only if its sub node didn't match.
+class AstNegateNode : public AstNode {
+ public:
+  AstNegateNode(AstNodePtr node) : node_{move(node)} {
+  }
+  virtual bool Check(std::string_view needle) const;
+  virtual std::string Debug() const;
+
+ private:
+  AstNodePtr node_;
+};
+
+// Ast logical operation node, matches only if sub nodes match
+// in respect to logical operation (and/or).
+class AstLogicalNode : public AstNode {
```

Comment:
Just an idea, feel free to dismiss.
Maybe define some `enum` within this class that tells it how to behave, instead of the current `bool disjunction`.
something like:
```
enum class Behavior {
  kAnd,
  kOr,
};
```

This should be clearer at the call site (`AstLogicalNode(..., kAnd)`), and is also more expandable (like we can add, in the future, `kXor`, `kNone`, `kAll`, etc).